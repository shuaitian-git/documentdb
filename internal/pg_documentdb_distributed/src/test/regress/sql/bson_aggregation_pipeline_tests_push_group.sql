SET search_path TO documentdb_core,documentdb_api,documentdb_api_catalog,documentdb_api_internal;

SET citus.next_shard_id TO 1012000;
SET documentdb.next_collection_id TO 10120;
SET documentdb.next_collection_index_id TO 10120;

SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 1, "product" : "beer", "pricingInfo" : { "msrp": 10, "retailPrice": 15 }, "stock" : 2, "year": 2020 }', NULL);
SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 2, "product" : "red wine", "pricingInfo" : { "msrp": 4, "retailPrice": 9 }, "stock" : 1, "year": 2021 }', NULL);
SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 3, "product" : "bread", "pricingInfo" : { "msrp": 3, "retailPrice": 11 }, "stock" : 5 , "year": 2020}', NULL);
SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 4, "product" : "whiskey", "pricingInfo" : { "msrp": 4, "retailPrice": 10 }, "stock" : 3 , "year": 2022}', NULL);
SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 5, "product" : "bread", "pricingInfo" : { "msrp": 75, "retailPrice": 100 }, "stock" : 1, "year": 2021 }', NULL);

/* running multiple $push accumulators with different expressions */
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "wholesalePricing":  { "$sum": ["$pricingInfo.msrp", 1] }, "qty": "$stock"} } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "retailPricing":  { "$subtract": ["$pricingInfo.retailPrice", 1] }, "isBread": { "$in": ["$product", ["bread"]] } } } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "shouldBeNull":  { "$subtract": ["$invalidName", 12] } } } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "combinedPrice":  { "$add": ["$pricingInfo.msrp", "$pricingInfo.retailPrice"] } } } } } ] }');

/* shard collection */
SELECT documentdb_api.shard_collection('db', 'testGroupWithPush', '{ "_id": "hashed" }', false);

/* run same $push queries to ensure consistency */
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "wholesalePricing":  { "$sum": ["$pricingInfo.msrp", 1] }, "qty": "$stock"} } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "retailPricing":  { "$subtract": ["$pricingInfo.retailPrice", 1] }, "isBread": { "$in": ["$product", ["bread"]] } } } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "shouldBeNull":  { "$subtract": ["$invalidName", 12] } } } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "combinedPrice":  { "$add": ["$pricingInfo.msrp", "$pricingInfo.retailPrice"] } } } } } ] }');

-- Test for missing values
SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 7, "product" : "bread", "pricingInfo" : { "msrp": 75, "retailPrice": 100 }, "year": 2021 }', NULL);

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": { "qty": "$stock"} } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": "$stock" } } } ] }');
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": ["$stock"] } } } ] }');
