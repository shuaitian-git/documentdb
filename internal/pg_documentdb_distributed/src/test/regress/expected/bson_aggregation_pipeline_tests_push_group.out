SET search_path TO documentdb_core,documentdb_api,documentdb_api_catalog,documentdb_api_internal;
SET citus.next_shard_id TO 1012000;
SET documentdb.next_collection_id TO 10120;
SET documentdb.next_collection_index_id TO 10120;
SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 1, "product" : "beer", "pricingInfo" : { "msrp": 10, "retailPrice": 15 }, "stock" : 2, "year": 2020 }', NULL);
NOTICE:  creating collection
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 2, "product" : "red wine", "pricingInfo" : { "msrp": 4, "retailPrice": 9 }, "stock" : 1, "year": 2021 }', NULL);
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 3, "product" : "bread", "pricingInfo" : { "msrp": 3, "retailPrice": 11 }, "stock" : 5 , "year": 2020}', NULL);
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT documentdb_api.insert_one('db','testGroupWithPush',' { "_id" : 4, "product" : "whiskey", "pricingInfo" : { "msrp": 4, "retailPrice": 10 }, "stock" : 3 , "year": 2022}', NULL);
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 5, "product" : "bread", "pricingInfo" : { "msrp": 75, "retailPrice": 100 }, "stock" : 1, "year": 2021 }', NULL);
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

/* running multiple $push accumulators with different expressions */
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "wholesalePricing":  { "$sum": ["$pricingInfo.msrp", 1] }, "qty": "$stock"} } } } ] }');
                                                                                                        document                                                                                                        
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "11" }, "qty" : { "$numberInt" : "2" } }, { "wholesalePricing" : { "$numberInt" : "4" }, "qty" : { "$numberInt" : "5" } } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "5" }, "qty" : { "$numberInt" : "1" } }, { "wholesalePricing" : { "$numberInt" : "76" }, "qty" : { "$numberInt" : "1" } } ] }
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "5" }, "qty" : { "$numberInt" : "3" } } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "retailPricing":  { "$subtract": ["$pricingInfo.retailPrice", 1] }, "isBread": { "$in": ["$product", ["bread"]] } } } } } ] }');
                                                                                        document                                                                                        
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "retailPricing" : { "$numberInt" : "14" }, "isBread" : false }, { "retailPricing" : { "$numberInt" : "10" }, "isBread" : true } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "retailPricing" : { "$numberInt" : "8" }, "isBread" : false }, { "retailPricing" : { "$numberInt" : "99" }, "isBread" : true } ] }
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "retailPricing" : { "$numberInt" : "9" }, "isBread" : false } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "shouldBeNull":  { "$subtract": ["$invalidName", 12] } } } } } ] }');
                                                 document                                                  
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "shouldBeNull" : null }, { "shouldBeNull" : null } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "shouldBeNull" : null }, { "shouldBeNull" : null } ] }
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "shouldBeNull" : null } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "combinedPrice":  { "$add": ["$pricingInfo.msrp", "$pricingInfo.retailPrice"] } } } } } ] }');
                                                                      document                                                                      
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "combinedPrice" : { "$numberInt" : "25" } }, { "combinedPrice" : { "$numberInt" : "14" } } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "combinedPrice" : { "$numberInt" : "13" } }, { "combinedPrice" : { "$numberInt" : "175" } } ] }
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "combinedPrice" : { "$numberInt" : "14" } } ] }
(3 rows)

/* shard collection */
SELECT documentdb_api.shard_collection('db', 'testGroupWithPush', '{ "_id": "hashed" }', false);
 shard_collection 
---------------------------------------------------------------------
 
(1 row)

/* run same $push queries to ensure consistency */
SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "wholesalePricing":  { "$sum": ["$pricingInfo.msrp", 1] }, "qty": "$stock"} } } } ] }');
                                                                                                        document                                                                                                        
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "5" }, "qty" : { "$numberInt" : "3" } } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "5" }, "qty" : { "$numberInt" : "1" } }, { "wholesalePricing" : { "$numberInt" : "76" }, "qty" : { "$numberInt" : "1" } } ] }
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "wholesalePricing" : { "$numberInt" : "4" }, "qty" : { "$numberInt" : "5" } }, { "wholesalePricing" : { "$numberInt" : "11" }, "qty" : { "$numberInt" : "2" } } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "retailPricing":  { "$subtract": ["$pricingInfo.retailPrice", 1] }, "isBread": { "$in": ["$product", ["bread"]] } } } } } ] }');
                                                                                        document                                                                                        
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "retailPricing" : { "$numberInt" : "9" }, "isBread" : false } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "retailPricing" : { "$numberInt" : "8" }, "isBread" : false }, { "retailPricing" : { "$numberInt" : "99" }, "isBread" : true } ] }
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "retailPricing" : { "$numberInt" : "10" }, "isBread" : true }, { "retailPricing" : { "$numberInt" : "14" }, "isBread" : false } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "shouldBeNull":  { "$subtract": ["$invalidName", 12] } } } } } ] }');
                                                 document                                                  
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "shouldBeNull" : null } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "shouldBeNull" : null }, { "shouldBeNull" : null } ] }
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "shouldBeNull" : null }, { "shouldBeNull" : null } ] }
(3 rows)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [ { "$group": { "_id": "$year", "items": { "$push": { "combinedPrice":  { "$add": ["$pricingInfo.msrp", "$pricingInfo.retailPrice"] } } } } } ] }');
                                                                      document                                                                      
---------------------------------------------------------------------
 { "_id" : { "$numberInt" : "2022" }, "items" : [ { "combinedPrice" : { "$numberInt" : "14" } } ] }
 { "_id" : { "$numberInt" : "2021" }, "items" : [ { "combinedPrice" : { "$numberInt" : "13" } }, { "combinedPrice" : { "$numberInt" : "175" } } ] }
 { "_id" : { "$numberInt" : "2020" }, "items" : [ { "combinedPrice" : { "$numberInt" : "14" } }, { "combinedPrice" : { "$numberInt" : "25" } } ] }
(3 rows)

-- Test for missing values
SELECT documentdb_api.insert_one('db','testGroupWithPush','{ "_id" : 7, "product" : "bread", "pricingInfo" : { "msrp": 75, "retailPrice": 100 }, "year": 2021 }', NULL);
                              insert_one                              
---------------------------------------------------------------------
 { "n" : { "$numberInt" : "1" }, "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": { "qty": "$stock"} } } } ] }');
                                                    document                                                     
---------------------------------------------------------------------
 { "_id" : "bread", "items" : [ { "qty" : { "$numberInt" : "5" } }, {  }, { "qty" : { "$numberInt" : "1" } } ] }
(1 row)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": "$stock" } } } ] }');
                                     document                                      
---------------------------------------------------------------------
 { "_id" : "bread", "items" : [ { "$numberInt" : "5" }, { "$numberInt" : "1" } ] }
(1 row)

SELECT document FROM documentdb_api_catalog.bson_aggregation_pipeline('db', '{ "aggregate": "testGroupWithPush", "pipeline": [{"$match": {"product": "bread"}}, { "$group": { "_id": "$product", "items": { "$push": ["$stock"] } } } ] }');
                                              document                                               
---------------------------------------------------------------------
 { "_id" : "bread", "items" : [ [ { "$numberInt" : "5" } ], [ null ], [ { "$numberInt" : "1" } ] ] }
(1 row)

